crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,599 @@
+use sui_types::transaction::GasData;
+use sui_types::transaction::ObjectArg;
+use sui_types::transaction::ProgrammableTransaction;
+use sui_types::transaction::TransactionData;
+use sui_types::transaction::TransactionDataAPI;
+use tap::Pipe;
+
+// TODO
+// - Updating the UnresolvedTransaction format to provide less information about inputs
+// - handle basic type inference and BCS serialization of pure args
+pub struct ResolveTransaction;
+
+impl ApiEndpoint<RestService> for ResolveTransaction {
+    fn method(&self) -> axum::http::Method {
+        axum::http::Method::POST
+    }
+
+    fn path(&self) -> &'static str {
+        "/transactions/resolve"
+    }
+
+    fn operation(
+        &self,
+        generator: &mut schemars::gen::SchemaGenerator,
+    ) -> openapiv3::v3_1::Operation {
+        OperationBuilder::new()
+            .tag("Transactions")
+            .operation_id("ResolveTransaction")
+            .query_parameters::<ResolveTransactionQueryParameters>(generator)
+            .request_body(
+                RequestBodyBuilder::new()
+                    // .json_content::<UnresolvedTransaction>(generator)
+                    .build(),
+            )
+            .response(
+                200,
+                ResponseBuilder::new()
+                    .json_content::<ResolveTransactionResponse>(generator)
+                    .bcs_content()
+                    .build(),
+            )
+            .build()
+    }
+
+    fn handler(&self) -> RouteHandler<RestService> {
+        RouteHandler::new(self.method(), resolve_transaction)
+    }
+}
+
+async fn resolve_transaction(
+    State(state): State<RestService>,
+    Query(parameters): Query<ResolveTransactionQueryParameters>,
+    accept: AcceptFormat,
+    Json(unresolved_transaction): Json<UnresolvedTransaction>,
+) -> Result<ResponseContent<ResolveTransactionResponse>> {
+    let executor = state
+        .executor
+        .as_ref()
+        .ok_or_else(|| anyhow::anyhow!("No Transaction Executor"))?;
+    let (reference_gas_price, protocol_config) = {
+        let system_state = state.reader.get_system_state_summary()?;
+
+        let current_protocol_version = state.reader.get_system_state_summary()?.protocol_version;
+
+        let protocol_config = ProtocolConfig::get_for_version_if_supported(
+            current_protocol_version.into(),
+            state.reader.inner().get_chain_identifier()?.chain(),
+        )
+        .ok_or_else(|| {
+            RestError::new(
+                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
+                "unable to get current protocol config",
+            )
+        })?;
+
+        (system_state.reference_gas_price, protocol_config)
+    };
+    let called_packages =
+        called_packages(&state.reader, &protocol_config, &unresolved_transaction)?;
+    let user_provided_budget = unresolved_transaction
+        .gas_payment
+        .as_ref()
+        .and_then(|payment| payment.budget);
```

Comment:
This needs to be before `resolved_transaction` due to passing `unresolved_transaction` by value in the `resolve_unresolved_transaction`?

Reply:
correct

Reply:
Thanks!